src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java

```diff
@@ -26,48 +28,106 @@
 public class BuildNameSetter extends BuildWrapper implements MatrixAggregatable {
 
     public final String template;
-
+    public final String matrixTemplate;
+    private boolean matches;
+    private int scenario = -1;
+    private boolean started = false;
+    
     @DataBoundConstructor
+    public BuildNameSetter(String template, String matrixTemplate) {
+        this.template = template;
+        this.matrixTemplate = matrixTemplate;
+        matches = template.equals(matrixTemplate);
+        
+        if(matches || matrixTemplate.length()>0 && template.length()==0){
+        	scenario = 0;
+        }else if (matrixTemplate.length()>0 && template.length()>0){
+        	scenario = 1;
+        }else if (matrixTemplate.length()==0 && template.length()>0){
+        	scenario = 2;
+        }
+    }
+    
+    @Deprecated
     public BuildNameSetter(String template) {
         this.template = template;
+        this.matrixTemplate = template;
     }
-
+    
     @Override
     public Environment setUp(AbstractBuild build, Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
-        setDisplayName(build, listener);
-
-        return new Environment() {
+        try{
+        	//not checking before setting name because this check is done via Jelly call to isMatrix()
+        	setDisplayName(build, listener);
+        }catch(Exception e){
+        	e.printStackTrace();
```

Comment:
Should be `e.printStackTrace(listener.getLogger())` otherwise the stack trace and error message are lost to standard out/err. Also the catch block should be changed to only catch `IllegalArgumentException`.